Userland/Libraries/LibAudio/MP3HuffmanTables.h

```diff
@@ -0,0 +1,1678 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ * Copyright (c) 2021, Arne Elster <arne@elster.li>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Array.h>
+#include <AK/BitStream.h>
+
+namespace Audio::MP3::Tables::Huffman {
+
+template<typename TSym>
+struct HuffmanEntry {
+    using symbol_type = TSym;
+    unsigned int code;
+    size_t code_size;
+    TSym symbol;
+};
+
+template<typename TSym>
+struct HuffmanNode {
+    int left;
+    int right;
+    size_t code_size;
+    TSym symbol;
+    bool is_leaf() const { return left == -1 && right == -1; }
+};
+
+template<typename TSym, size_t N>
+using HuffmanEntries = AK::Array<TSym, N>;
+
+template<typename TSym, size_t N>
+using HuffmanNodes = AK::Array<HuffmanNode<TSym>, N>;
+
+template<typename TSym, size_t N>
+constexpr size_t MaxHuffmanTableLength(const HuffmanEntries<TSym, N>& table)
+{
+    size_t max = 0;
+    for (size_t i = 1; i < table.size(); i++) {
+        if (table[i].code_size > max)
+            max = table[i].code_size;
+    }
+    return max;
+}
+
+template<typename TSym, size_t N>
+constexpr int _MakeHuffmanTree(const auto& table, HuffmanNodes<TSym, N>&& nodes)
+{
+    int allocationCount = 1;
+
+    nodes[0].left = nodes[0].right = -1;
+
+    for (size_t i = 0; i < table.size(); i++) {
+        const HuffmanEntry<TSym>& entry = table[i];
+        int tree_pointer = 0;
+
+        for (size_t j = 0; j < entry.code_size; j++) {
+            const bool bit = (entry.code >> (entry.code_size - j - 1)) & 1;
+            const bool eoc = j == entry.code_size - 1;
+
+            int& target = bit ? nodes[tree_pointer].left : nodes[tree_pointer].right;
+
+            if (target != -1) {
+                if (eoc)
+                    return -1;
+                tree_pointer = target;
+            } else {
+                tree_pointer = target = allocationCount++;
+                nodes[target].left = nodes[target].right = -1;
+                nodes[target].code_size = j + 1;
+                if (eoc)
+                    nodes[target].symbol = entry.symbol;
+            }
+        }
+    }
+
+    return allocationCount;
+}
+
+template<auto table>
+constexpr auto MakeHuffmanTree()
+{
+    using TSym = decltype(table)::value_type::symbol_type;
+    constexpr int size_worst_case = 1 << (MaxHuffmanTableLength(table) + 1);
+    constexpr int size_fitted = _MakeHuffmanTree<TSym>(table, HuffmanNodes<TSym, size_worst_case>());
+    static_assert(size_fitted >= 0, "Huffman tree is ill-formed");
+
+    HuffmanNodes<TSym, size_fitted> nodes;
+    _MakeHuffmanTree<TSym>(table, move(nodes));
+
+    return nodes;
+}
+
+template<typename T>
+const HuffmanNode<T>* huffman_decode(InputBitStream& bitstream, const HuffmanNode<T>* root)
+{
+    const HuffmanNode<T>* node = root;
+
+    while (!node->is_leaf() && !bitstream.has_any_error()) {
```

Comment:
I changed it so that it returns an `Optional`. A bitstream error doesn't lead to undefined behaviour as all read bits will simply be 0. The bitstream error can then be handled further up the loader.